examples/ticks/date_index_formatter.py

```diff
@@ -0,0 +1,80 @@
+"""
+=====================================
+Custom tick formatter for time series
+=====================================
+
+.. redirect-from:: /gallery/text_labels_and_annotations/date_index_formatter
+.. redirect-from:: /gallery/ticks/date_index_formatter2
+
+When plotting daily data, e.g., financial time series, one often wants
+to leave out days on which there is no data, for instance weekends, so that
+the data are plotted at regular intervals without extra spaces for the days
+with no data.
+The example shows how to use an 'index formatter' to achieve the desired plot.
+"""
+
+import numpy as np
+import matplotlib.pyplot as plt
+import matplotlib.cbook as cbook
+import matplotlib.lines as ml
+from matplotlib.dates import DateFormatter, DayLocator
+from matplotlib.ticker import Formatter
+
+
+# Load a numpy record array from yahoo csv data with fields date, open, high,
+# low, close, volume, adj_close from the mpl-data/sample_data directory. The
+# record array stores the date as an np.datetime64 with a day unit ('D') in
+# the date column.
+r = (cbook.get_sample_data('goog.npz', np_load=True)['price_data']
+     .view(np.recarray))
+r = r[:9]  # get the first 9 days
+
+fig, (ax1, ax2) = plt.subplots(nrows=2, figsize=(6, 6),
+                               constrained_layout={'hspace': .15})
+
+# First we'll do it the default way, with gaps on weekends
+ax1.plot(r.date, r.adj_close, 'o-')
+
+# Highlight gaps in daily data
+gaps = np.flatnonzero(np.diff(r.date) > np.timedelta64(1, 'D'))
+for gap in r[['date', 'adj_close']][np.stack((gaps, gaps + 1)).T]:
+    ax1.plot(gap.date, gap.adj_close, 'w--', lw=2)
+ax1.legend(handles=[ml.Line2D([], [], ls='--', label='Gaps in daily data')])
+
+ax1.set_title("Plot y at x Coordinates")
+ax1.xaxis.set_major_locator(DayLocator())
+ax1.xaxis.set_major_formatter(DateFormatter('%a'))
+
+
+# Next we'll write a custom index formatter
+def format_date(x, _):
+    try:
+        return r.date[round(x)].item().strftime('%a')
+    except IndexError:
+        pass
+
+# Create an index plot (x defaults to range(len(y)) if omitted)
+ax2.plot(r.adj_close, 'o-')
```

Comment:
Maybe just be explicit rather than make the user go back and parse the comment? 
```suggestion
# Create an index plot:
ax2.plot(np.arange(0, len(r.adj_close)), r.adj_close, 'o-')
```

Reply:
I made this change on purpose to underline/explain that this is an index plot which by definition plots the values versus an index number. In my view, the plot command is more legible/clearer without the range definition as the first argument. Besides, this kind of plot is explicitly advertised in [`plot`](https://matplotlib.org/stable/api/_as_gen/matplotlib.axes.Axes.plot.html#matplotlib.axes.Axes.plot), so I'd rather keep it as is.

Reply:
I think, however, that most of the core devs consider this a nasty holdover from the Matlab days, and prefer the more explicit interface.  Certainly when I was reading this I was confused about what `r.adj_close` was, and how it was plotting x and y, and then I had to go up, figure out what `r.adj_close` was and then figure out why it didn't need two arguments.  

In general, I wonder if it is more confusing than not to keep `r` as a structured array?  How common do we think those are in the wild (I never use them, but then I use xarray all the time).

Reply:
> most of the core devs consider this a nasty holdover from the Matlab days, and prefer the more explicit interface

hm, I find this feature quite handy, but if this usage is not recommended it should be described as discouraged in the docs, shouldn't it

> I wonder if it is more confusing than not to keep r as a structured array

Alternatives (pandas, xarray, ...) require additional dependencies, csv requires additional date conversion, so why not leave it as a structured array (given that the data structure is described in detail in the comment at lines 24-27)?